x/ibc/testing/chain.go

```diff
@@ -0,0 +1,373 @@
+package testing
+
+import (
+	"fmt"
+	"testing"
+	"time"
+
+	"github.com/stretchr/testify/require"
+	abci "github.com/tendermint/tendermint/abci/types"
+	"github.com/tendermint/tendermint/crypto"
+	"github.com/tendermint/tendermint/crypto/secp256k1"
+	lite "github.com/tendermint/tendermint/lite2"
+	tmtypes "github.com/tendermint/tendermint/types"
+
+	"github.com/cosmos/cosmos-sdk/simapp"
+	sdk "github.com/cosmos/cosmos-sdk/types"
+	authtypes "github.com/cosmos/cosmos-sdk/x/auth/types"
+	connectiontypes "github.com/cosmos/cosmos-sdk/x/ibc/03-connection/types"
+	channeltypes "github.com/cosmos/cosmos-sdk/x/ibc/04-channel/types"
+	ibctmtypes "github.com/cosmos/cosmos-sdk/x/ibc/07-tendermint/types"
+	commitmenttypes "github.com/cosmos/cosmos-sdk/x/ibc/23-commitment/types"
+	host "github.com/cosmos/cosmos-sdk/x/ibc/24-host"
+	"github.com/cosmos/cosmos-sdk/x/ibc/keeper"
+)
+
+const (
+	// Default params used to create a TM client
+	TrustingPeriod  time.Duration = time.Hour * 24 * 7 * 2
+	UnbondingPeriod time.Duration = time.Hour * 24 * 7 * 3
+	MaxClockDrift   time.Duration = time.Second * 10
+
+	ConnectionVersion = "1.0"
+	ChannelVersion    = "1.0"
+
+	ConnectionIDPrefix = "connectionID"
+	ChannelIDPrefix    = "channelID"
+	PortIDPrefix       = "portID"
+)
+
+// TestChain is a testing struct that wraps a simapp with the last TM Header, the current ABCI
+// header and the validators of the TestChain. It also contains a field called ClientID. This
+// is the clientID that *other* chains use to refer to this TestChain. For simplicity's sake
+// it is also the chainID on the TestChain Header. The SenderAccount is used for delivering
+// transactions through the application state.
+type TestChain struct {
+	t *testing.T
+
+	ClientID      string
```

Comment:
should this be an array? in case a chain has both a localhost and tendermint clients defined for example? Or would one rather create a new `TestChain` instance for that?

Reply:
I think it makes sense to test having multiple clients on a chain. Do we want to autogenerate clientID names like done for connections/channels/ports?

Reply:
> Do we want to autogenerate clientID names like done for connections/channels/ports?

Yeah that makes sense 👍